src/lustre/lustreTypeChecker.ml

```diff
@@ -242,8 +242,229 @@ let extract_consts: LA.const_clocked_typed_decl -> tc_context
+  | LA.TernaryOp (pos, top, e1, e2, e3) -> eval_bool_ternary_op ctx pos top e1 e2 e3
+  | LA.CompOp (pos, cop, e1, e2) -> eval_comp_op ctx pos cop e1 e2
+  | e -> type_error (LH.pos_of_expr e) ("Cannot evaluate expression" ^ string_of_expr e)  
+(** try and evalutate expression to bool, return error otherwise *)
+
+and eval_bool_binary_op: tc_context -> Lib.position -> LA.binary_operator -> LA.expr -> LA.expr -> bool tc_result = 
+  fun ctx pos bop e1 e2 ->
+  eval_bool_expr ctx e1 >>= fun v1 ->
+  eval_bool_expr ctx e2 >>= fun v2 ->
+  match bop with
+  | And -> R.ok (v1 && v2) 
+  | Or -> R.ok (v1 || v2)
+  | Xor -> R.ok ((v1 && not v2) || (v2 && not v1))
+  | Impl -> R.ok (not v1 || v2)
+  | _ -> type_error pos ("Cannot evaluate non-bool binary expression"
+                         ^ string_of_expr (LA.BinaryOp (pos, bop, e1, e2))
+                         ^" to a bool value")
+(** try and evalutate binary op expression to bool, return error otherwise *)
   
+and eval_bool_ternary_op: tc_context -> Lib.position -> LA.ternary_operator
+                     -> LA.expr -> LA.expr -> LA.expr -> bool tc_result
+  = fun ctx pos top b1 e1 e2 ->
+  eval_bool_expr ctx b1 >>= fun c ->
+  eval_bool_expr ctx e1 >>= fun v1 ->
+  eval_bool_expr ctx e2 >>= fun v2 ->
+  match top with
+  | LA.Ite -> if c then R.ok v1 else R.ok v2
+  | LA.With -> type_error pos "With operator is not supported"
+(** try and evalutate ternary op expression to bool, return error otherwise *)
+
+and eval_int_ternary_op: tc_context -> Lib.position -> LA.ternary_operator
+                     -> LA.expr -> LA.expr -> LA.expr -> int tc_result
+  = fun ctx pos top b1 e1 e2 ->
+  match top with
+  | LA.Ite ->
+     eval_bool_expr ctx b1 >>= fun c ->
+     if c
+     then eval_int_expr ctx e1
+     else eval_int_expr ctx e2
+  | LA.With -> type_error pos "With operator is not supported"
+(** try and evalutate ternary op expression to int, return error otherwise *)
+
+             
+and eval_comp_op: tc_context -> Lib.position -> LA.comparison_operator -> LA.expr -> LA.expr -> bool tc_result = 
+  fun ctx pos cop e1 e2 ->
+  eval_int_expr ctx e1 >>= fun v1 ->
+  eval_int_expr ctx e2 >>= fun v2 ->
+  match cop with
+  | Eq -> R.ok (v1 = v2)
+  | Neq -> R.ok (v1 <> v2)
+  | Lte -> R.ok (v1 <= v2)
+  | Lt -> R.ok (v1 < v2)
+  | Gte -> R.ok (v1 > v2)
+  | Gt -> R.ok (v1 >= v2)
+(** try and evalutate comparison op expression to bool, return error otherwise *)
+
+and simplify_array_index: tc_context -> Lib.position -> LA.expr -> LA.expr -> LA.expr
+  = fun ctx pos e1 idx -> 
+  match (simplify_expr ctx e1) with
+  | LA.GroupExpr (pos, ArrayExpr, es) ->
+     (match (eval_int_expr ctx idx) with
+      | Ok i -> if List.length es > i
+                then List.nth es i
+                else
+                  (Log.log L_fatal "Trying to access element out of bounds %a"
+                     Lib.pp_print_position pos
+                  ; raise OutOfBounds)
```

Comment:
Debug messages are the only kind of message this module should log, not error messages. You should catch this exception within this module (`type_check_program`), and return an `Error`.

Reply:
The problem is identifying if evaluation fails due to a sub-term being a symbolic constant or if it fails as it is an 
ill formed term is assigned the same `Error _`. However we need to have different behaviour outcomes. 
For previous case it should be an expression and for the later it is an evaluation error. 
eg. 
```
arr = false^10
z = arr[50]
```

This is a well typed term but the evaluation should fail.
however,

```
const c: int;
arr = false^c
z = arr[50]
```
is well typed, but the evaluation should not evaluate.